### ISSUE TITLE
Change encoding to utf-8 for reading .ini files

### ISSUE THREAD

**Issue by ilg-ul** [mesonbuild/meson#9996]
**Describe the bug**

The code reading .ini files assumes the content to be ASCII, and fails if the files include utf-8 characters.

```console
> meson setup --backend ninja --buildtype debug -D platform-name=native --native-file xpacks/micro-os-plus-build-helper/meson/toolchains/gcc.ini --native-file tests/platform-native/meson/native.ini build/native-meson-gcc11-debug tests
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/mesonbuild/mesonmain.py", line 228, in run
    return options.run_func(options)
  File "/usr/local/lib/python3.6/dist-packages/mesonbuild/msetup.py", line 281, in run
    app.generate()
  File "/usr/local/lib/python3.6/dist-packages/mesonbuild/msetup.py", line 177, in generate
    env = environment.Environment(self.source_dir, self.build_dir, self.options)
  File "/usr/local/lib/python3.6/dist-packages/mesonbuild/environment.py", line 522, in __init__
    config = coredata.parse_machine_files(self.coredata.config_files)
  File "/usr/local/lib/python3.6/dist-packages/mesonbuild/coredata.py", line 934, in parse_machine_files
    parser = MachineFileParser(filenames)
  File "/usr/local/lib/python3.6/dist-packages/mesonbuild/coredata.py", line 880, in __init__
    self.parser.read(filenames)
  File "/usr/lib/python3.6/configparser.py", line 697, in read
    self._read(fp, filename)
  File "/usr/lib/python3.6/configparser.py", line 1015, in _read
    for lineno, line in enumerate(fp, start=1):
  File "/usr/lib/python3.6/encodings/ascii.py", line 26, in decode
    return codecs.ascii_decode(input, self.errors)[0]
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc2 in position 29: ordinal not in range(128)
```

The culprit is a `µ` character used in the header comment present in the .ini files, like:

```ini
#
# This file is part of the µOS++ distribution. <--- utf-8 characters!
#   (https://github.com/micro-os-plus)
# Copyright (c) 2022 Liviu Ionescu
#
# Permission to use, copy, modify, and/or distribute this software
# for any purpose is hereby granted, under the terms of the MIT license.
#
# If a copy of the license was not distributed with this file, it can
# be obtained from https://opensource.org/licenses/MIT/.
#
# -----------------------------------------------------------------------------

# https://mesonbuild.com/Machine-files.html

[built-in options]
warning_level = '0'
werror = false
layout = 'flat' # 'mirror'

# -----------------------------------------------------------------------------
```

**To Reproduce**

Use a comment line with the **µ** character in the .ini files.

**Expected behavior**

The .ini files be parsed with utf-8 encodding, not ascii.

**system parameters**
* the build happens inside an Ubuntu 18.04 docker container, with minimal content
* Python 3.6.9
* 0.59.4

I tried to inject the encoding via the environment `PYTHONIOENCODING=utf-8`, but did not help.

Unfortunately my knowledge of Python is not good enough to find a solution and suggest a pull request.

As a temporary workaround I removed the utf-8 characters from the .ini files, and the builds passed.

**Comment by ilg-ul** [comment]
And a related issue from the same family:

```
Traceback (most recent call last):
  File "/Host/home/ilg/Work/meson-build-0.59.4-1/linux-arm/install/meson-build/lib/python3.9/mesonbuild/mesonmain.py", line 228, in run
  File "/Host/home/ilg/Work/meson-build-0.59.4-1/linux-arm/install/meson-build/lib/python3.9/mesonbuild/mtest.py", line 1998, in run
  File "/Host/home/ilg/Work/meson-build-0.59.4-1/linux-arm/install/meson-build/lib/python3.9/mesonbuild/mtest.py", line 1678, in doit
  File "/Host/home/ilg/Work/meson-build-0.59.4-1/linux-arm/install/meson-build/lib/python3.9/mesonbuild/mtest.py", line 1817, in run_tests
  File "/Host/home/ilg/Work/meson-build-0.59.4-1/linux-arm/install/meson-build/lib/python3.9/asyncio/base_events.py", line 642, in run_until_complete
  File "/Host/home/ilg/Work/meson-build-0.59.4-1/linux-arm/install/meson-build/lib/python3.9/mesonbuild/mtest.py", line 1920, in _run_tests
  File "/Host/home/ilg/Work/meson-build-0.59.4-1/linux-arm/install/meson-build/lib/python3.9/mesonbuild/mtest.py", line 683, in finish
  File "/Host/home/ilg/Work/meson-build-0.59.4-1/linux-arm/install/meson-build/lib/python3.9/mesonbuild/mtest.py", line 588, in report_progress
  File "/Host/home/ilg/Work/meson-build-0.59.4-1/linux-arm/install/meson-build/lib/python3.9/mesonbuild/mtest.py", line 558, in emit_progress
  File "/Host/home/ilg/Work/meson-build-0.59.4-1/linux-arm/install/meson-build/lib/python3.9/mesonbuild/mtest.py", line 522, in print_progress
UnicodeEncodeError: 'ascii' codec can't encode character '\U0001f311' in position 8: ordinal not in range(128)
```

This time the issue is caused by the test output which is displayed as:

```
diag::trace test started

  Check trace::initialize
    \u2713 initial count
    \u2713 initial content x
    \u2713 count initialised
    \u2713 initial content cleared

  Check trace::putchar
    \u2713 count increased by 1
    \u2713 buffer has c
```

The output should be:

```
diag::trace test started
 
   Check trace::initialize
     ✓ initial count
     ✓ initial content x
     ✓ count initialised
     ✓ initial content cleared
 
   Check trace::putchar
     ✓ count increased by 1
     ✓ buffer has c
```

**Comment by eli-schwartz** [comment]
> The code reading .ini files assumes the content to be ASCII

Not really true. The code reading .ini files doesn't assume the content at all, that is the entire problem.

Rather what happens here is that it uses the default locale, which in your case is apparently not UTF-8. Also, you are using Python 3.6 (which Meson no longer supports in git master).

> I tried to inject the encoding via the environment `PYTHONIOENCODING=utf-8`, but did not help.

Regrettably, PYTHONIOENCODING (a highly python-specific variable) only affects standard input, standard output, and standard error. It doesn't affect any other kinds of data, such as files.

The good news is that setting LC_CTYPE (a universal variable from the LC_ALL family of subset variables, that affects all programs) changes the locale for everything, including both stdin/stdout/stderr as well as reading/writing files or processing the data included in them.

The other good news is https://www.python.org/dev/peps/pep-0538/ which specifies for Python 3.7 and onward that if your LC_CTYPE is not UTF-8 enabled, Python will generally insist on upgrading it, so this problem therefore goes away.

**Comment by eli-schwartz** [comment]
I am not sure what to do about this issue. Meson dropped support for Python 3.6, so all supported versions of Python work fine. On the other hand, the most recent stable release of Meson still supports Python 3.6. On the other other hand, Python 3.6 is no longer supported by the Python Software Foundation (this is why we dropped support in Meson git master).

And it's not really clear what to do to solve this problem anyway. The simplest solution is to simply guarantee that the locale is set. I'm not sure how to do a thorough vetting of Meson code and ensure that any function which opens a file or even passes a file to the stdlib to open it handles encodings, especially when this is only for maintenance releases and not git master.

**Comment by ilg-ul** [comment]
> Meson dropped support for Python 3.6

Not directly related. The issue is present with Python 3.9 too, just that the test that I showed you was on an older system.

**Comment by ilg-ul** [comment]
> And it's not really clear what to do to solve this problem anyway. The simplest solution is to simply guarantee that the locale is set. 

How do you guarantee this?

If the environment variables are not present, can't you adjust the environment at run time to add them? Or set some Python process wide defaults to UTF-8?

Breaking with a stack trace because an environment variable is not set is hardly an acceptable behaviour.

>  I'm not sure how to do a thorough vetting of Meson code and ensure that any function which opens a file or even passes a file to the stdlib to open it handles encodings, especially when this is only for maintenance releases and not git master.

Unfortunately I don't have an answer to this.

All I know is that I tried to run my tests on a minimalistic Ubuntu 20 container, and meson crashed. I identified the culprit, I removed the wide characters from the .ini files, then meson did not properly display the test output.

At the same time, exactly the same tests, in the same environment, but built and run with CMake, were fine. 

Such a run is still available on GitHub, with the meson tests at the end:

- <https://github.com/micro-os-plus/diag-trace-xpack/runs/5167376929>

My projects have dual-build configurations, and I test them both with CMake and meson, so it is quite easy to compare the two. The CMake syntax may be unusual, but in the end it does the job fairly well. The meson syntax may seem easier to read, but the countless restrictions make things way more difficult, and it is not guaranteed that will result in a functional build.

For my tests I spent some time to experiment with various docker images and I was finally able to find one with an environment that no longer breaks meson, but this is not how a professional tool is expected to behave.

If you have no solution to this, feel free to close the issue at any time.

**Comment by jpakkane** [comment]
There are a bunch of other places where we explicitly set the encoding to utf-8. The change would be to change `self.parser.read(filenames)` to `self.parser.read(filenames, encoding='utf'8')`.

The only question is would this break something? Since the current implementation fails on anything non-ascii, probably not.

**Comment by eli-schwartz** [comment]
Again, this works fine on python 3.10 on my machine. Because it's a locale issue and ever since python 3.7, python itself will force your locale from ASCII to utf-8.

**Comment by ilg-ul** [comment]
> ever since python 3.7, python itself will force your locale from ASCII to utf-8

I cannot confirm this, I first encountered this bug with the python 3.9 that I embedded in my distribution, but, in order to give you a reproducible case, and avoid giving you a reason to shift the blame to my distribution, I retested with the system python.

Python may have its share of issues, but by defensive programming style guides, this is a bug in your application, it is not normal to crash with a stack trace.

If you don't consider fixing it, at least catch the exception and display a civilised message informing the user how to properly set the locales to make it work.

**Comment by eli-schwartz** [comment]
For the record, I am a dummy and totally wrong, and I apologize for confusing the issue.

[PEP 538](https://www.python.org/dev/peps/pep-0538/) enables locale coercion.
[PEP 540](https://www.python.org/dev/peps/pep-0540/) enables a UTF-8 mode in file opening/reading.

Both are available in Python 3.7, and the former is the default, but the latter is NOT a default, even though it arguably should be.

[PEP 597](https://www.python.org/dev/peps/pep-0597/) enables a way to detect PEP 540 issues in your codebase.

It is only available in Python 3.10, though.

Since despite locale coercion the default encoding for `open()` is still a non-unicode one, we would need to somehow manually opt into this. We cannot ask users to do this, especially since it's a pretty niche python-specific variable. We cannot programmatically set it, because AFAIK python doesn't let you modify this.

What we *can* do is, of course, insert the hack of "manually specify sane encoding", which indeed `run_custom_lint.py` tries to check for, but the thing is, it does it *badly*. It's a string grep for the text `open(` (and a couple other things) and it misfires and complains about lots of innocent things, and it completely overlooks cases like configparser where an innocent function accepts an encoding kwarg which it internally uses when internally calling `open()`.

> I'm not sure how to do a thorough vetting of Meson code and ensure that any function which opens a file or even passes a file to the stdlib to open it handles encodings

I figured out a way to do this, which is much more reliable than a string grep.

**Comment by eli-schwartz** [comment]
And my new method of finding issues caught an additional case not reported by this ticket. :D
